File: RoRServer/src/main/java/models/session/RoRSession.java
package models.session;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import communication.MessageInformation;
import communication.queue.receiver.QueueReceiver;
import exceptions.MapNotFoundException;
import models.base.ModelBase;
import models.game.Map;
import models.game.Player;
import persistent.MapManager;

/**
 * Oberklasse von EditorSession und GameSession
 * 
 */
public abstract class RoRSession extends ModelBase {
	private ArrayList<Player> players = new ArrayList<>();
	private Map map;
	protected boolean started;
	private String mapName;

	protected QueueReceiver queueReceiver;
	
	public RoRSession(String name) {
		super(name);
	}
		
	public void setup() {
		queueReceiver.setup();
	}

	public void removePlayer(Player player) {
		this.players.remove(player);
		notifyPlayerLeft(player);
	}

	public void removePlayers() {
		for (int i = getPlayers().size() - 1; i >= 0; i--) {
			Player player = getPlayers().get(i);
			removePlayer(player);
		}
	}

	protected abstract void notifyPlayerLeft(Player player);

	public Player getHost() {
		return players.stream().filter(x -> x.getIsHost()).findFirst().get();
	}

	public Map getMap() {
		return map;
	}

	public void setMap(Map map) {
		this.map = map;
	}

	public List<Player> getPlayers() {
		return Collections.unmodifiableList(players);
	}
	
	public void addPlayer(Player player) {
		players.add(player);
	}
	
	public Player getPlayerById(UUID id) {
		for (Player player : getPlayers()) {
			if (player.getId().equals(id))
				return player;
		}
		return null;
	}

	public boolean isStarted() {
		return started;
	}

	public void start() {
		started = true;
		MessageInformation messageInfo = new MessageInformation("Start");
		notifyChange(messageInfo);
	}

	public String getMapName() {
		return mapName;
	}

	public void setMapName(String mapName) {
		this.mapName = mapName;
		notifyChangedMapName();
	}

	/**
	 * Schickt eine Message mit dem neuen MapName, über den Topic der GameSession,
	 * an alle angemeldeten Clients 
	 */
	private void notifyChangedMapName() {
		MessageInformation messageInfo = new MessageInformation("ChangeMapSelection");
		messageInfo.putValue("mapName", getMapName());
		if(this instanceof GameSession) {
			GameSession gameSession = (GameSession) this;	
			int availablePlayers = MapManager.getAvailablePlayerSlots(getMapName());
			gameSession.setAvailablePlayerSlots(availablePlayers);
			messageInfo.putValue("availablePlayerSlots", gameSession.getAvailablePlayerSlots());
		
		}
		notifyChange(messageInfo);
	}
}

File: RoRClient/Models/Session/RoRSession.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RoRClient.Models.Game;
using System.Collections.ObjectModel;
using RoRClient.Communication.Queue;
using RoRClient.Models.Base;
using RoRClient.Communication.Topic;
using RoRClient.Communication.DataTransferObject;
using RoRClient.Communication;

namespace RoRClient.Models.Session
{
    public class RoRSession : ObservableBase
    {
        protected string name;
        protected Map map;
	    protected string mapName;

		protected ObservableCollection<Player> players = new ObservableCollection<Player>();

        protected QueueSender queueSender;
        protected TopicReceiver topicReceiver;

        public RoRSession()
        {
            map = new Map();
        }

        public string Name
        {
            get
            {
                return name;
            }
            set
            {
                name = value;
            }
        }

        public void Init(string topicName)
        {
            queueSender = new QueueSender(topicName);
        }


        public Map Map
        {
            get
            {
                return map;
            }
            set
            {
                map = value;
            }
        }

        public ObservableCollection<Player> Players
        {
            get
            {
                return players;
            }
        }

        public Player OwnPlayer
        {
            get
            {
                Player ownPlayer = players.Where(x => x.Id == ClientConnection.GetInstance().ClientId).FirstOrDefault();
                if (ownPlayer != null)
                {
                    return ownPlayer;
                }
                return new Player(Guid.NewGuid(), "");
            }
        }

        public void AddPlayer(Player player)
        {
            players.Add(player);
        }

        public void RemovePlayer(Player player)
        {
            players.Remove(player);
        }

        public Player GetPlayerById(Guid playerId)
        {
            return players.Where(x => x.Id == playerId).FirstOrDefault();
        }
        public QueueSender QueueSender
        {
            get
            {
                return queueSender;
            }
        }

	    public string MapName
	    {
		    get { return mapName; }
		    set
		    {
			    if (mapName != value)
			    {
				    mapName = value;
				    NotifyPropertyChanged("MapName");
			    }
		    }
	    }

	    private bool started;
	    public bool Started
	    {
		    get
		    {
			    return started;
		    }
		    set
		    {
			    if (started != value)
			    {
				    started = value;
				    NotifyPropertyChanged("Started");
			    }
		    }
	    }
	}
}
